rgw: fix consistency bug with OLH objects

cfsnyder opened the pull request:
Relates to: https://tracker.ceph.com/issues/61359

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

cbodley commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -5471,7 +5471,7 @@ static void generate_fake_tag(const DoutPrefixProvider *dpp, RGWRados* store, ma
 static bool is_olh(map<string, bufferlist>& attrs)
 {
-  map<string, bufferlist>::iterator iter = attrs.find(RGW_ATTR_OLH_INFO);
+  map<string, bufferlist>::iterator iter = attrs.find(RGW_ATTR_OLH_VER);
```
this part makes sense. `RGW_ATTR_OLH_VER` is set during `olh_init_modification()`, while `RGW_ATTR_OLH_INFO` is added later by `apply_olh_log()`. so for `get_obj_state()` calls in between, this prevents us from misinterpreting the olh head object as a normal head object. with this change, we'll go on to call `get_olh_target_state()` and `follow_olh()` to look up the current version. i just want to make sure that this call path returns `-ENOENT` until `apply_olh_log()` completes

currently, `follow_olh()` will return `-EINVAL` if it doesn't find the `RGW_ATTR_OLH_INFO` attr: https://github.com/ceph/ceph/blob/1b5e1bf/src/rgw/driver/rados/rgw_rados.cc#L7794-L7796

should that return `-ENOENT` now? there's a similar check in `RGWRados::get_olh()` that should probably change too

cfsnyder replied:
I think that is a good point and sounds like you're probably correct. I'll review further to confirm as well.

cfsnyder replied:
I did confirm this and updated it now

cbodley commented on qa/workunits/rgw/test_rgw_reshard.py:
```diff
@@ -313,13 +315,22 @@ def boto_connect(portnum, ssl, proto):
+    # TESTCASE 'check that PUT succeeds during reshard'
+    log.debug(' test: PUT succeeds during reshard')
+    num_shards = get_bucket_stats(VER_BUCKET_NAME).num_shards
+    exec_cmd('''radosgw-admin --rgw-debug-reshard-completion-delay 5 \
+                      bucket reshard --bucket {} --num-shards {}'''
```
these reshard tests already make heavy use of a `FaultInjector` class to test error paths and crash recovery. in https://github.com/ceph/ceph/pull/51733, i just added the ability to inject delays as well

that would allow this test case to use `--inject-delay-at=do_reshard --inject-delay-ms=5000` without any changes to the reshard code itself

cfsnyder replied:
@cbodley that makes sense and is a neat concept. I'm a bit confused by the statement about this being usable "without any changes to the reshard code itself", though? Is there some dynamic nature to this fault injector code that I'm missing or do I still need to add a fault injector hook in `do_reshard` to trigger this delay at the appropriate place?

cbodley replied:
injecting a sleep at the existing 'do_reshard' check (https://github.com/ceph/ceph/blob/3bc1ea5/src/rgw/driver/rados/rgw_reshard.cc#L972-L974) should behave the same as the sleep you added inside of `do_reshard()`. the bucket index is locked for resharding over all of this, so the racing `put_object()` call would be blocked either way

cfsnyder replied:
Ah, I'd missed that there was already an existing 'do_reshard' check. I've included your commits and updated this to  use the FaultInjector now. It seems to be working as advertised.

cfsnyder commented:
> I still want to investigate and add test cases for the scenario where an object transitions from unversioned to versioned, because I suspect that these changes have the potential to cause problems there

I added https://github.com/ceph/ceph/pull/51700/files#diff-fb30258ed43305d97a0fdeb02fdb91bc0e1613985e4dfbeab46381c4f1719bc4R6818-R6829 to handle that case properly and added a test case for it

cbodley commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -7477,28 +7530,78 @@ int RGWRados::apply_olh_log(const DoutPrefixProvider *dpp,
+
+  r = rgw_rados_operate(dpp, ref.pool.ioctx(), ref.obj.oid, &rm_op, null_yield);
+  if (r == -ECANCELED) {
+    return r; /* someone else made a modification in the meantime */
+  } else {
```
nit: remove the `else` after `return` to save a level of nesting for the rest of the function

cfsnyder commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -6743,6 +6743,49 @@ int RGWRados::obj_operate(const DoutPrefixProvider *dpp, const RGWBucketInfo& bu
   return rgw_rados_operate(dpp, ref.pool.ioctx(), ref.obj.oid, op, &outbl, null_yield);
 }
 
+int RGWRados::olh_cancel_modification(const DoutPrefixProvider *dpp, const RGWBucketInfo& bucket_info, RGWObjState& state, const rgw_obj& olh_obj, const std::string& op_tag)
+{
+  if (cct->_conf->rgw_debug_inject_olh_cancel_modification_err) {
+    // simulate the scenario where we fail to remove the pending xattr
+    return -cct->_conf->rgw_debug_inject_olh_cancel_modification_err;
```
the configurable is really there so the tests can simulate what was happening before these changes (olh cancelation/cleanup not happening on request threads). It's particularly useful for generating the sort of bucket index garbage that we need for verifying the functionality of the new radosgw-admin commands. I can change the return type to void and log errors within this function vs. in the calling function though.

cfsnyder commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -6743,6 +6743,49 @@ int RGWRados::obj_operate(const DoutPrefixProvider *dpp, const RGWBucketInfo& bu
+  ObjectWriteOperation op;
+  bucket_index_guard_olh_op(dpp, state, op);
+  op.rmxattr(attr_name.c_str());
+  r = rgw_rados_operate(dpp, ref.pool.ioctx(), ref.obj.oid, &op, null_yield);
```
will do

cfsnyder commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -7562,15 +7664,24 @@ int RGWRados::set_olh(const DoutPrefixProvider *dpp, RGWObjectCtx& obj_ctx,
-    ret = bucket_index_link_olh(dpp, bucket_info, *state, target_obj,
-				delete_marker, op_tag, meta, olh_epoch, unmod_since,
-				high_precision_time, y, zones_trace, log_data_change);
+    if (cct->_conf->rgw_debug_inject_set_olh_err) {
+      // fail here to simulate the scenario of an unlinked object instance
+      ret = -cct->_conf->rgw_debug_inject_set_olh_err;
+    } else {
+      ret = bucket_index_link_olh(dpp, bucket_info, *state, target_obj,
```
this PR fixes the issues with the double delete case so we do need some way to inject an error to reliably reproduce  now

cbodley commented on src/rgw/rgw_admin.cc:
```diff
@@ -1402,6 +1411,42 @@ int set_bucket_quota(rgw::sal::Driver* driver, OPT opt_cmd,
+  listable = false;
+  int ret;
+  do {
+    ret = store->bi_list(bs, key.name, marker, 1000, &entries, &is_truncated);
```
consider using `cls_rgw_bucket_list_op()` instead, which should only be returning the plain entries. `bi_list()` would search each of the special namespaces too

cfsnyder replied:
I'm breaking out of the loop when it encounters a non-plain entry, so it won't actually iterate over the special namespaces. The issue with using `cls_rgw_bucket_list_op` is that it's susceptible to high latencies and errors when a bunch of these bogus entries are in the index (the ones that we're trying to clean up). That said, I acknowledge that I do need to do something different here because breaking out of the loop when I hit the first non-plain entry means that I'm not considering any plain entries that come after the special namespaces. I'll review this.

cbodley reviewed (commented):
high-level feedback for the admin commands:
* this adds too much to rgw_admin.cc. please move the logic to driver/rados/rgw_bucket.* similar to check_index(). please create separate functions for "check olh" vs "check unlinked" instead of branching internally
* instead of just logging the number of matches, can we dump a json array with all of their keys? i believe that's how 'bucket check' works
* please consider using coroutines instead of threads. a coroutine function that handles a single shard would be much easier to read and maintain. if you want to limit how many shards we process at a time, spawn N coroutines that each process a different value of (shard % N)

cfsnyder commented:
> high-level feedback for the admin commands:
> 
>     * this adds too much to rgw_admin.cc. please move the logic to driver/rados/rgw_bucket.* similar to check_index(). please create separate functions for "check olh" vs "check unlinked" instead of branching internally

Ok I agree, I will refactor accordingly.

> 
>     * instead of just logging the number of matches, can we dump a json array with all of their keys? i believe that's how 'bucket check' works

Yep that sounds good.

> 
>     * please consider using coroutines instead of threads. a coroutine function that handles a single shard would be much easier to read and maintain. if you want to limit how many shards we process at a time, spawn N coroutines that each process a different value of (shard % N)

I did consider that initially, but since rgw_admin doesn't currently use coroutines or set up an asio io_context, I wasn't sure if it made sense to include in a bugfix PR. I'll give it a try and see what it looks like.

cfsnyder commented on src/rgw/rgw_admin.cc:
```diff
@@ -8286,6 +8353,189 @@ int main(int argc, const char **argv)
+            return;
+          }
+          RGWRados::BucketShard bs(store);
+          string marker = opt_cmd == OPT::BUCKET_CHECK_OLH ? "\x80" "1001_" : "\x80" "1000_";
```
Yeah I know. I don't feel great about that either, but it's a significant optimization to skip listing all of the irrelevant parts of the keyspace and I didn't think it would be a great idea to add new cls methods just for these specialty listing cases.

cbodley commented:
> I'll give it a try and see what it looks like.

you can find an example in rgw_sync_checkpoint.cc `rgw_bucket_sync_checkpoint()` for `BUCKET_SYNC_CHECKPOINT`

cfsnyder commented:
@cbodley 

> instead of just logging the number of matches, can we dump a json array with all of their keys? i believe that's how 'bucket check' works

I'm questioning whether this is a good idea now that I've got around to implementing it. I'm currently outputting progress information via stderr logs and then the JSON output goes to stdout. When dumping all of the keys in a JSON array, the command output with no redirection ends up looking like this:

```
2023-06-20T09:34:46.444+0000 7f4491b96f40  1 NOTICE: finished shard 0 (0 entries found)
2023-06-20T09:34:46.444+0000 7f4491b96f40  1 NOTICE: finished shard 1 (0 entries found)
2023-06-20T09:34:46.448+0000 7f4491b96f40  1 NOTICE: finished shard 2 (0 entries found)
2023-06-20T09:34:46.452+0000 7f4491b96f40  1 NOTICE: finished shard 3 (0 entries found)
2023-06-20T09:34:46.452+0000 7f4491b96f40  1 NOTICE: finished shard 4 (0 entries found)
[
    "d"2023-06-20T09:34:46.456+0000 7f4491b96f40  1 NOTICE: finished shard 5 (1 entries found)
2023-06-20T09:34:46.460+0000 7f4491b96f40  1 NOTICE: finished shard 6 (0 entries found)
2023-06-20T09:34:46.460+0000 7f4491b96f40  1 NOTICE: finished shard 7 (0 entries found)
,
    "c"2023-06-20T09:34:46.464+0000 7f4491b96f40  1 NOTICE: finished shard 8 (1 entries found)
2023-06-20T09:34:46.468+0000 7f4491b96f40  1 NOTICE: finished shard 9 (0 entries found)

]
2023-06-20T09:34:46.468+0000 7f4491b96f40  1 NOTICE: finished shard 10 (0 entries found)
2023-06-20T09:34:46.468+0000 7f4491b96f40  1 NOTICE: finished all shards (2 entries found)
```

That obviously isn't very nice for the user experience. Unfortunately, we have cases where there are millions of bogus entries per shard, so it isn't practical to buffer everything in memory and only print all of the keys to stdout at the end.

Maybe I should add another cli flag so that a file path can be specified for output of the keys? What are your thoughts?

cbodley commented:
> That obviously isn't very nice for the user experience. Unfortunately, we have cases where there are millions of bogus entries per shard, so it isn't practical to buffer everything in memory and only print all of the keys to stdout at the end.

are the per-shard messages really necessary? to avoid excessive buffering, you could keep a counter and call `formatter->flush(std::cout)` regularly

cbodley commented:
the merge of https://github.com/ceph/ceph/pull/50206 introduced several conflicts in rgw_rados.*. hopefully that will make it easier to get real `optional_yield`s everywhere

cbodley commented:
i'm eager to get this ready for reef, so i scheduled an early round of testing in https://pulpito.ceph.com/cbodley-2023-06-23_19:56:03-rgw-pr-51700-distro-default-smithi/

cbodley commented:
results in https://pulpito.ceph.com/cbodley-2023-06-23_19:56:03-rgw-pr-51700-distro-default-smithi/ look pretty good 👍

cbodley commented:
@cfsnyder with the reef release looming, i'd like to prioritize the merge/backport of the bug fixes and test cases. if necessary, i think it's okay for the repair tooling to miss the release and land in the first reef point release. as long as the bug fixes are in reef, that wouldn't prevent us from backporting everything to quincy then pacific

would you mind moving the radosgw-admin commands to a separate PR so we can get the rest onto reef asap? is there anything else missing from the fixes or test cases? afaik those parts are good to merge

cfsnyder commented:
@cbodley yeah I can move the radosgw-admin commands to a different PR. I'm not aware of anything that is missing from the fixes or test cases.

cfsnyder commented:
jenkins test make check

cbodley commented:
i've been unable to get successful builds for this in https://shaman.ceph.com/builds/ceph/pr-51700/, trying again..

cbodley commented:
[testing](https://pulpito.ceph.com/cbodley-2023-07-11_16:38:08-rgw:verify-main-distro-default-smithi/) with the `versioning.yaml` rename is consistently failing here:
```
2023-07-11T18:57:35.658 INFO:tasks.workunit.client.0.smithi101.stdout:connected to http://localhost:80
2023-07-11T18:57:35.665 INFO:tasks.workunit.client.0.smithi101.stderr:Traceback (most recent call last):
2023-07-11T18:57:35.665 INFO:tasks.workunit.client.0.smithi101.stderr:  File "/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_versioning.py", line 106, in <module>
2023-07-11T18:57:35.666 INFO:tasks.workunit.client.0.smithi101.stderr:    main()
2023-07-11T18:57:35.666 INFO:tasks.workunit.client.0.smithi101.stderr:  File "/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_versioning.py", line 79, in main
2023-07-11T18:57:35.666 INFO:tasks.workunit.client.0.smithi101.stderr:    json_out = json.loads(out)
2023-07-11T18:57:35.666 INFO:tasks.workunit.client.0.smithi101.stderr:  File "/usr/lib64/python3.6/json/__init__.py", line 349, in loads
2023-07-11T18:57:35.667 INFO:tasks.workunit.client.0.smithi101.stderr:    s = s.decode(detect_encoding(s), 'surrogatepass')
2023-07-11T18:57:35.667 INFO:tasks.workunit.client.0.smithi101.stderr:UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 1954: invalid start byte
2023-07-11T18:57:35.705 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

cfsnyder commented:
> [testing](https://pulpito.ceph.com/cbodley-2023-07-11_16:38:08-rgw:verify-main-distro-default-smithi/) with the `versioning.yaml` rename is consistently failing here:
> 
> ```
> 2023-07-11T18:57:35.658 INFO:tasks.workunit.client.0.smithi101.stdout:connected to http://localhost:80
> 2023-07-11T18:57:35.665 INFO:tasks.workunit.client.0.smithi101.stderr:Traceback (most recent call last):
> 2023-07-11T18:57:35.665 INFO:tasks.workunit.client.0.smithi101.stderr:  File "/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_versioning.py", line 106, in <module>
> 2023-07-11T18:57:35.666 INFO:tasks.workunit.client.0.smithi101.stderr:    main()
> 2023-07-11T18:57:35.666 INFO:tasks.workunit.client.0.smithi101.stderr:  File "/home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/test_rgw_versioning.py", line 79, in main
> 2023-07-11T18:57:35.666 INFO:tasks.workunit.client.0.smithi101.stderr:    json_out = json.loads(out)
> 2023-07-11T18:57:35.666 INFO:tasks.workunit.client.0.smithi101.stderr:  File "/usr/lib64/python3.6/json/__init__.py", line 349, in loads
> 2023-07-11T18:57:35.667 INFO:tasks.workunit.client.0.smithi101.stderr:    s = s.decode(detect_encoding(s), 'surrogatepass')
> 2023-07-11T18:57:35.667 INFO:tasks.workunit.client.0.smithi101.stderr:UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 1954: invalid start byte
> 2023-07-11T18:57:35.705 DEBUG:teuthology.orchestra.run:got remote process result: 1
> ```

I'll look into this today. Tested briefly this morning and see that it is still passing for me with a vstart cluster.

cfsnyder commented on qa/suites/rgw/verify/tasks/versioning.yml:
```diff
@@ -0,0 +1,5 @@
+tasks:
+- workunit:
+    clients:
+      client.0:
+        - rgw/run-versioning.sh
```
yep I'll update that

cfsnyder commented on qa/workunits/rgw/test_rgw_versioning.py:
```diff
@@ -0,0 +1,107 @@
+    log.debug('TEST: verify that index entries and OLH objects are cleaned up after index linking error\n')
+    key = str(uuid.uuid4())
+    try:
+        exec_cmd('ceph config set client.rgw rgw_debug_inject_set_olh_err 2')
```
Wondering if I need a brief sleep between this command and the next one. Seems like it's possible that this config update just wasn't applied prior to the following delete call when run in teuthology.

cbodley replied:
using the running radosgw's admin socket to change config would probably be more reliable, but i'm okay with the sleeps if they pass consistently

cfsnyder replied:
I'm questioning whether it's really related to the config being applied now, but I'll update the tests again to use the admin socket and will remove the special chars that cause json decoding to fail so that the we fail on the following assertion instead. Any other recommendations for how to debug these failures since they aren't happening with a vstart cluster @cbodley ?

cbodley replied:
the use of `client.rgw` might be problematic. vstart uses entity names like `client.rgw.8000`, but in these teuthology jobs the name is `client.0`

cbodley replied:
to work in both environments, i think we need a command-line option or environment variable to pass in the right entity name or admin socket path

cbodley replied:
i scheduled a job in https://pulpito.ceph.com/cbodley-2023-07-13_17:57:16-rgw:verify-main-distro-default-smithi/ that replaces `client.rgw` with `client.0`

cbodley replied:
that didn't pick up the right suite-branch, rerun in https://pulpito.ceph.com/cbodley-2023-07-13_18:29:03-rgw:verify-main-distro-default-smithi/

cfsnyder replied:
is there an easy way to get the name of the rgw client that works in both contexts or should i just set the config at the `client` or `global` level instead?

cbodley replied:
http://qa-proxy.ceph.com/teuthology/cbodley-2023-07-13_18:29:03-rgw:verify-main-distro-default-smithi/7336231/teuthology.log passed with `config set client.0`

> should i just set the config at the client or global level instead?

i would expect `client` to work in all cases, yeah

cfsnyder replied:
updated

cbodley commented:
rescheduled the rgw/verify suite with existing builds against your updated suite-branch in https://pulpito.ceph.com/cbodley-2023-07-12_19:51:14-rgw:verify-main-distro-default-smithi

cbodley commented:
same failures again

cfsnyder commented:
jenkins test signed

cbodley commented:
latest version passed a single job: https://pulpito.ceph.com/cbodley-2023-07-14_14:13:10-rgw:verify-main-distro-default-smithi/

scheduled the whole suite 🤞 https://pulpito.ceph.com/cbodley-2023-07-14_16:57:08-rgw-main-distro-default-smithi/

cbodley commented:
lots of dead jobs due to lab issues. scheduled a --rerun in https://pulpito.ceph.com/cbodley-2023-07-17_14:52:30-rgw-main-distro-default-smithi

cbodley commented:
i don't think that's due to anything in this pr, it was just an unlikely race between the `reshard add`/`reshard list` commands and radosgw's resharding thread. note the timestamps from [teuthology.log](http://qa-proxy.ceph.com/teuthology/cbodley-2023-07-17_14:52:30-rgw-main-distro-default-smithi/7341683/teuthology.log):
```
2023-07-17T16:33:59.989 INFO:tasks.workunit.client.0.smithi105.stderr:radosgw-admin reshard add --bucket a-bucket --num-shards 2
2023-07-17T16:34:00.025 INFO:tasks.workunit.client.0.smithi105.stderr:ignoring --setuser ceph since I am not root
2023-07-17T16:34:00.025 INFO:tasks.workunit.client.0.smithi105.stderr:ignoring --setgroup ceph since I am not root
...
2023-07-17T16:34:00.153 INFO:tasks.workunit.client.0.smithi105.stderr:2023-07-17T16:34:00.130+0000 7f73b1615f40 20 remove_watcher() i=7
2023-07-17T16:34:00.165 INFO:tasks.workunit.client.0.smithi105.stderr:radosgw-admin reshard list
```

and the reshard logging from [rgw.ceph.client.0.log](http://qa-proxy.ceph.com/teuthology/cbodley-2023-07-17_14:52:30-rgw-main-distro-default-smithi/7341683/remote/smithi105/log/rgw.ceph.client.0.log.gz):
```
2023-07-17T16:34:00.122+0000 7f845fc16700 20 rgw reshard worker thread: process_entry resharding a-bucket
...
2023-07-17T16:34:00.142+0000 7f845fc16700  1 rgw reshard worker thread: execute INFO: reshard of bucket "a-bucket" completed successfully
```
i scheduled a re-rerun in https://pulpito.ceph.com/cbodley-2023-07-17_20:43:19-rgw-main-distro-default-smithi to see if it goes away

cbodley commented:
lab issues causing too many dead jobs. i'll try again tomorrow

cbodley commented:
down to one dead job in https://pulpito.ceph.com/cbodley-2023-07-17_21:21:12-rgw-main-distro-default-smithi/, and the failures all look unrelated. good enough for me!

cbodley commented:
@cfsnyder i closed https://tracker.ceph.com/issues/59663 as resolved, and moved https://tracker.ceph.com/issues/61359 to Pending Backport. could you please prepare the reef backport asap so i can schedule tests?

separately, we'll follow up on the radosgw-admin command for cleanup. can you please create a new tracker issue for that part?

cfsnyder commented:
@cbodley thanks, yeah I'll work on the backport this morning

cfsnyder commented:
Added this tracker for the radosgw-admin commands: https://tracker.ceph.com/issues/62075
